File: src/QuickSort.java
class Solution {
    public int[] sortArray(int[] nums) {
        int len = nums.length;
        int l = 0, r = len-1;

        quicksort(nums, l, r);
        return nums;
    }

    public void quicksort(int[] nums, int l, int r) {

        if (l >= r) return;

        int pivot = handlePivot(nums, l, r);
        int guard = nums[r];

        int nextswap = l;
        for (int i = l; i < r; i++) {
            if (nums[i] <= guard) {
                int t = nums[nextswap];
                nums[nextswap] = nums[i];
                nums[i] = t;
                nextswap++;
            }
        }
        nums[r] = nums[nextswap];
        nums[nextswap] = guard;

        quicksort(nums, nextswap + 1, r);
        quicksort(nums, l, nextswap-1);
    }

    public void swap(int[] nums, int i1, int i2) {
        int t = nums[i1];
        nums[i1] = nums[i2];
        nums[i2] = t;
    }

    public int handlePivot(int[] nums, int l, int r) {
        int m = l + (r - l)/2;

        if (nums[l] > nums[m]) swap(nums, l, m);
        if (nums[l] > nums[r]) swap(nums, l, r);
        if (nums[m] < nums[r]) swap(nums, m, r);
   }
}

File: src/QuickSort.cpp

#include "stdafx.h"
#include <iostream>
#include "timer.h"

using namespace std;

/*
rm ./quick.out ; g++ QuickSort.cpp -o quick.out && ./quick.out
*/
// Array based
class Solution {
public:

    static void printA(int *a, int n) {
        for (int i = 0; i < n; i++) {
            cout<<a[i] << " ";
        }
        cout << "\n";
    }

    static void sortArray(int* a, int n, int& switchTimes)
    {
        quickSort(a, 0, n-1);
    }

    static void quickSort(int *a, int s, int e)
    {
        //cout <<"s ,e = " << s << ", " << e << "\n";
        if (s >= e) return;

        int m = partition(a, s, e);

        quickSort(a, s, m-1);
        quickSort(a, m+1, e);
    }

    static int partition(int *a, int l, int r) 
    {
        int guard = a[r];
        int m = l;

        for (int i = l; i < r; i++) {
            if (a[i] <= guard) {
                int n = a[i];
                a[i] = a[m];
                a[m++] = n;
            }
        }
        a[r] = a[m];
        a[m] = guard;

        return m;
    }

};

int main() {
	{
    Timer t("quick");

    for (int i = 0; i < 5000; i++) 
    {
        {
            int switchTimes;
            int a[10] = {4, 1, 9, 5, 2, 3, 8, 1, 0, 7};
            Solution::sortArray(a, 10, switchTimes);
        }

        {
            int switchTimes;
            int a[10] = {9,8,7,6,5,4,3,2,1,0};
            Solution::sortArray(a, 10, switchTimes);
        }
        {
            int switchTimes;
            int a[10] = {0,1,2,3,4,5,10,7,8,9};
            Solution::sortArray(a, 10, switchTimes);
        }
    }
    }
	system("pause");
}

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
